### PR TITLE
PIMOB-1416: Add UI Test for full UI journey

### PR DESCRIPTION
**Pull request by alex-ioja-yang-cko**
## Proposed changes

[Jira task](https://checkout.atlassian.net/browse/PIMOB-1416)

## Types of changes

What types of changes does your code introduce to Frames Ios?
_Put an `x` in the boxes that apply_

* [ ] Bugfix (non-breaking change which fixes an issue)
* [X] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

* [X] Lint and unit tests pass locally with my changes
* [X] I have added tests that prove my fix is effective or that my feature works
* [ ] I have added necessary documentation (if appropriate)

## Further comments

Ended up covering a few minor fixes that came out from the UI test.

### CODE REVIEW

File: Source/UI/PaymentForm/View/BillingFormSummaryView.swift
Comment (alex-ioja-yang-cko) [inline Source/UI/PaymentForm/View/BillingFormSummaryView.swift:62]: This method already gets called with correct style passed  on line 65 so this was doing an extra refresh with wrong style

File: Source/UI/BillingForm/View/BillingFormHeaderCell.swift
Comment (ehab-al-cko) [inline Source/UI/BillingForm/View/BillingFormHeaderCell.swift:19]: please share some details why need to inject `startEnabled` f or common Button View?

File: Source/UI/BillingForm/View/BillingFormHeaderCell.swift
Comment (ehab-al-cko) [inline Source/UI/BillingForm/View/BillingFormHeaderCell.swift:55]: please share details for deleting these delegates?

File: Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift
Comment (ehab-al-cko) [inline Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift:203]: this is duplicate from another PR. I think, you can share same logic to the other PR and i will test it.

File: bitrise.yml
Comment (ehab-al-cko) [inline bitrise.yml:302]: I think, main device is iPhone 13. 
```suggestion
    UITEST_SIMULATOR_DEVICE: iPhone 13
```

File: iOS Example Frame/iOS Example Frame/Factory/Factory+UITest.swift
Comment (ehab-al-cko) [inline iOS Example Frame/iOS Example Frame/Factory/Factory+UITest.swift:31]: please move `iOS Example Frame/iOS Example Frame/Factory/Factory+UITest.swift` to UITest target

File: iOS Example Frame/iOS Example Frame/Factory/Factory+UITest.swift
Comment (ehab-al-cko) [inline iOS Example Frame/iOS Example Frame/Factory/Factory+UITest.swift:93]: Is there UI test for current default screen without duplicating the code?

File: iOS Example Frame/iOS Example Frame/Factory/Factory.swift
Comment (ehab-al-cko) [inline iOS Example Frame/iOS Example Frame/Factory/Factory.swift:61]: this is code is for merchants. it should not include UITest logic

**Review by ehab-al-cko — CHANGES_REQUESTED** [review]
Please add more details how to test it because it's not working with me

File: Source/UI/BillingForm/View/BillingFormHeaderCell.swift
Comment (alex-ioja-yang-cko) [reply Source/UI/BillingForm/View/BillingFormHeaderCell.swift:19]: The button starts by default as enabled. Some buttons (like Pay and Billing Done) should start as disabled and only be enabled once validation passes.

This may not be visible to us in normal testing due to the speed of the action, but it becomes easily flagged by UI Test

File: Source/UI/BillingForm/View/BillingFormHeaderCell.swift
Comment (alex-ioja-yang-cko) [reply Source/UI/BillingForm/View/BillingFormHeaderCell.swift:55]: This is dead code which was not called anywhere.

File: Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift
Comment (alex-ioja-yang-cko) [reply Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift:203]: Actually this is what testing is for, to flag issues in the code. And I think this PR did well to catch the boolean difference, as well as the fact the Country update itself was not triggering a validation recheck.

Also as mentioned in the other PR, the solution proposed there is to use inefficient and verbose syntax, not to actually fix a simple boolean comparison.

File: bitrise.yml
Comment (alex-ioja-yang-cko) [reply bitrise.yml:302]: I am not sure what "main device" refers to and what impact this proposed change has?

I chose the Pro version on the naive assumption that "was the simulator going to provide improve improved performance (as real device would claim), our simulator may theoretically run the tests faster". I don't think this is a valid assumption, but either way, no harm to use the pro version...

What is your motivation for not using the Pro version?

File: iOS Example Frame/iOS Example Frame/Factory/Factory.swift
Comment (alex-ioja-yang-cko) [reply iOS Example Frame/iOS Example Frame/Factory/Factory.swift:61]: This is a demo app which comes with UITests. 
This is not code for merchants as it does not form the SDK consumer would use, its built on top of it.

File: iOS Example Frame/iOS Example Frame/Factory/Factory+UITest.swift
Comment (alex-ioja-yang-cko) [reply iOS Example Frame/iOS Example Frame/Factory/Factory+UITest.swift:93]: If we were to do UI tests without specific UI customisations designed for the UI tests we would end up with one of the following:
- demo designs that are not what UX wants
- UI tests that do not test for "edgy" behaviours

This can be improved by creating our demo designs to match UX templates and also contain our edge case scenarios, but then it may extend the number of buttons, making it harder to digest on first sight and would make the UI tests flakier as their UI itself may become subject to change on UX whim. 
Using the specific scheme with unique template overrides ensures the UI test is separated from the demo UI we present to clients.

File: iOS Example Frame/iOS Example Frame/Factory/Factory+UITest.swift
Comment (alex-ioja-yang-cko) [reply iOS Example Frame/iOS Example Frame/Factory/Factory+UITest.swift:31]: The call site for the API should be able to discover the declarations of that API.

In practice that means if we were to declare the methods contained within this file to UITest target, our application would need to have access to the UITest module. That would be a wrong dependency, as UITest needs to depend on source code it tests, not the other way around.

File: Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift
Comment (alex-ioja-yang-cko) [inline Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift:178]: @ehab-al-cko please leave comments that require some response on a code comment, not on final review, so we can actually discuss.

You said
> Please add more details how to test it because it's not working with me

Could you please describe what exactly does that mean? The UI tests are passing on pipeline, so it appears to work on a 3rd party machine.

File: Source/UI/BillingForm/View/BillingFormHeaderCell.swift
Comment (ehab-al-cko) [reply Source/UI/BillingForm/View/BillingFormHeaderCell.swift:19]: UI Test **must not** change normal common UI component. please undo it.

File: Source/UI/BillingForm/View/BillingFormHeaderCell.swift
Comment (ehab-al-cko) [reply Source/UI/BillingForm/View/BillingFormHeaderCell.swift:55]: How come? if you did a little search you will find it in 
```
extension BillingFormViewController: BillingFormHeaderCellDelegate {

    func doneButtonIsPressed() {
        delegate?.doneButtonIsPressed(sender: self)
    }

    func cancelButtonIsPressed() {
        delegate?.cancelButtonIsPressed(sender: self)
    }
}```

File: Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift
Comment (ehab-al-cko) [reply Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift:203]: Let's wait for other PR to be merged then this part will be solved.

File: bitrise.yml
Comment (ehab-al-cko) [reply bitrise.yml:302]: I think, we need to agree on with device should run UI test

File: iOS Example Frame/iOS Example Frame/Factory/Factory+UITest.swift
Comment (ehab-al-cko) [reply iOS Example Frame/iOS Example Frame/Factory/Factory+UITest.swift:93]: I think, one of ACs to test default screen as a normal user. 
[- _re create the same UI test with new UI flow_](https://checkout.atlassian.net/browse/PIMOB-1360)

File: iOS Example Frame/iOS Example Frame/Factory/Factory.swift
Comment (ehab-al-cko) [reply iOS Example Frame/iOS Example Frame/Factory/Factory.swift:61]: It's an Example app and they will reuse code. We may discuss this with team in next standup

File: Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift
Comment (ehab-al-cko) [reply Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift:178]: Like steps to  run UITest. It's not working with normal scheme.

File: Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift
Comment (deepesh-vasthimal-cko) [reply Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift:203]: @alex-ioja-yang-cko As i understand this small change might conflict with other PR what @ehab-al-cko has already worked on, So would be worth discussing the changes before or waiting for other PRs to be merged. Will save time and effort.

File: Source/UI/BillingForm/View/BillingFormHeaderCell.swift
Comment (alex-ioja-yang-cko) [reply Source/UI/BillingForm/View/BillingFormHeaderCell.swift:55]: Please see the code you referred to and the code being deleted

File: Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift
Comment (alex-ioja-yang-cko) [reply Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift:178]: If its not working with normal scheme, could you please try to use UITest scheme instead? UITests are normally a separate target and as such commonly they require a separate scheme to run.

File: Source/UI/BillingForm/View/BillingFormHeaderCell.swift
Comment (alex-ioja-yang-cko) [reply Source/UI/BillingForm/View/BillingFormHeaderCell.swift:19]: Not sure if you read my message properly. 
The current behaviour is buggy, enabling button when it should be disabled. 
The tests did what any tests do, catch bugs.
To get the test passing, code level changes were done.

Please don't make this harder, take a minute to understand replies and ask if unclear. Don't demand things without understanding the basics behind.

Comment (ehab-al-cko) [reply Source/UI/BillingForm/View/BillingFormHeaderCell.swift:19]: I think you misunderstand me. 
 if `ButtonView` has a new bug then it should be unit tested not a UI test.  `startEnabled ` is confusing name. you may change it to `isEnabled` like native apple button. Also, remove `isEnabled` from `ElementButtonStyle` if you want to inject button disable state.
```
public protocol ElementButtonStyle: ElementStyle {
    var isEnabled: Bool { get set }
```
Please undo changes here and a new PR to fix this issue with unit test would be great because this change will affect behaviour of buttons and needs to be tested alone.

File: Source/UI/BillingForm/View/BillingFormHeaderCell.swift
Comment (ehab-al-cko) [reply Source/UI/BillingForm/View/BillingFormHeaderCell.swift:55]: ??

Comment (alex-ioja-yang-cko) [reply Source/UI/BillingForm/View/BillingFormHeaderCell.swift:55]: The code being deleted is `doneAction` and `cancelAction`. You're looking at different things

File: Source/UI/BillingForm/View/BillingFormHeaderCell.swift
Comment (alex-ioja-yang-cko) [reply Source/UI/BillingForm/View/BillingFormHeaderCell.swift:19]: Ok, so different problems to address here.

 

## Testing

Unit tests and UITests are both **tests** (same as snapshots, integration etc).
The general role for testing is to find bugs, enforce behaviours and prevent unintended changes.
Different kind of tests find bugs in different ways, for example:
- **unit tests** are great at handling business logic issues, inconsistent states, parsings, data flows, etc
- **snapshot tests** help identify UI level changes and ensure with some degree of certainty the layout, positioning, formatting of the UI, etc
- **UI tests** don't do either of these. They check the usability of the UI, navigation, presence ...

> if ButtonView has a new bug then it should be unit tested not a UI test.

As I have already stated 
_Some buttons (like Pay and Billing Done) should start as disabled and only be enabled once validation passes_ 
    and 
_This may not be visible to us in normal testing due to the speed of the action, but it becomes easily flagged by UI Test_

ButtonView does not have a bug as far as Unit Tests are concerned. Furthermore ButtonView is an UIView type and Unit Testing UI components is one of the reasons MVVM exists (**to not have to**).

 

## Inject button state

The way I see these components being designed to work, they get told their configuration (via a Style interface with multiple inheritances) what their state should be.

The bug is in the way ButtonView is being enabled very early in the lifecycle of the ViewController. This is due to its default state and the delay from UI being shown to the state being updated. 

I don't think affecting that update flow is part of the solution for this problem. The solution is to be able to control the start state, which is prior to the style becoming available.

So please, if you have a preference to a different approach feel free to detail it and as always I am happy to try if technically sound.

File: bitrise.yml
Comment (alex-ioja-yang-cko) [reply bitrise.yml:302]: Agreement starts with discussion. From my side I have proposed "iPhone 13 Pro" in the PR and as you have raised the subject I have asked you to have your say.

Please let me know when you'd like to touch this complex subject.

File: iOS Example Frame/iOS Example Frame/Factory/Factory+UITest.swift
Comment (alex-ioja-yang-cko) [reply iOS Example Frame/iOS Example Frame/Factory/Factory+UITest.swift:93]: Could you please justify why you think so? And what does "default screen as a normal user" actually mean to you, because I am unsure we have the same understanding for that.

**Review by ehab-al-cko — APPROVED** [review]

File: Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift
Comment (ehab-al-cko) [reply Source/UI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift:203]: i think, if this PR is merged then no need for other PR.

**Review by deepesh-vasthimal-cko — APPROVED** [review]

**Pull request merged**
